### PR TITLE
Add a simple-to-understand example for online DPO

### PR DESCRIPTION
**Pull request by vwxyzjn**
This PR adds a simple-to-understand example for online DPO

### CODE REVIEW

**Comment by qgallouedec** [comment]
🔥

File: docs/source/online_dpo_trainer.md [hunk 1]
```diff
@@ -7,6 +7,41 @@ While [Guo et al. (2024)](https://huggingface.co/papers/2402.04792) used a LLM j
...
+```python
+from datasets import Dataset
+from trl import OnlineDPOConfig, OnlineDPOTrainer
+from transformers import (
+    AutoModelForCausalLM,
+    AutoModelForSequenceClassification,
+    AutoTokenizer,
+)
+NUM_DUMMY_SAMPLES = 100
+tok = AutoTokenizer.from_pretrained("gpt2")
```
Comment (lewtun) [inline docs/source/online_dpo_trainer.md:21]: turbo nit:
```suggestion
tokenizer = AutoTokenizer.from_pretrained("gpt2")
```

File: docs/source/online_dpo_trainer.md [hunk 2]
```diff
@@ -7,6 +7,41 @@ While [Guo et al. (2024)](https://huggingface.co/papers/2402.04792) used a LLM j
...
+tok.add_special_tokens({"pad_token": "[PAD]"})
```
Comment (lewtun) [inline docs/source/online_dpo_trainer.md:22]: Is this a quirk of GPT2? Is so, WDYT about using a different model like one of the Qwen2 or SmolLM ones?

File: docs/source/online_dpo_trainer.md [hunk 3]
```diff
@@ -7,6 +7,41 @@ While [Guo et al. (2024)](https://huggingface.co/papers/2402.04792) used a LLM j
...
+model = AutoModelForCausalLM.from_pretrained("gpt2")
+ref_model = AutoModelForCausalLM.from_pretrained("gpt2")
+reward_model = AutoModelForSequenceClassification.from_pretrained("gpt2", num_labels=1)
```
Comment (lewtun) [inline docs/source/online_dpo_trainer.md:25]:
```suggestion
# The model to score completions with. In practice, you will need a fine-tuned reward model. See Reward Bench for some good ones: https://huggingface.co/spaces/allenai/reward-bench
reward_model = AutoModelForSequenceClassification.from_pretrained("gpt2", num_labels=1)
```

File: docs/source/online_dpo_trainer.md [hunk 4]
```diff
@@ -7,6 +7,41 @@ While [Guo et al. (2024)](https://huggingface.co/papers/2402.04792) used a LLM j
...
+train_dataset = Dataset.from_dict(
+    {"input_ids": [tok.encode("Q: Hi how are you? A:")] * NUM_DUMMY_SAMPLES})
+eval_dataset = Dataset.from_dict(
+    {"input_ids": [tok.encode("Q: What do you like to eat A:")] * NUM_DUMMY_SAMPLES})
+trainer = OnlineDPOTrainer(
+    OnlineDPOConfig(
+        output_dir="models/minimal/demo",
```
Comment (lewtun) [inline docs/source/online_dpo_trainer.md:32]: turbo nit (I don't think we need the nested dir):
```suggestion
        output_dir="online-dpo-model",
```

File: docs/source/online_dpo_trainer.md [hunk 5]
```diff
@@ -7,6 +7,41 @@ While [Guo et al. (2024)](https://huggingface.co/papers/2402.04792) used a LLM j
...
+from trl import OnlineDPOConfig, OnlineDPOTrainer
+from transformers import (
+    AutoModelForCausalLM,
+    AutoModelForSequenceClassification,
+    AutoTokenizer,
+)
+NUM_DUMMY_SAMPLES = 100
+tok = AutoTokenizer.from_pretrained("gpt2")
+tok.add_special_tokens({"pad_token": "[PAD]"})
+model = AutoModelForCausalLM.from_pretrained("gpt2")
```
Comment (lewtun) [inline docs/source/online_dpo_trainer.md:23]:
```suggestion
# The model to optimise
model = AutoModelForCausalLM.from_pretrained("gpt2")
```

**Review by lewtun — APPROVED** [review]
Thanks a lot Costa! LGTM with some nits :)

File: docs/source/online_dpo_trainer.md (hunk 2 above)
Comment (vwxyzjn) [reply docs/source/online_dpo_trainer.md:22]: Let's use SmolLM

**Comment by vwxyzjn** [comment]
image


Training also works with the dummy example
